executor/aggfuncs/func_count_distinct.go

```diff
@@ -433,3 +435,386 @@ func appendJSON(encodedBytes, _ []byte, val json.BinaryJSON) []byte {
 	encodedBytes = append(encodedBytes, val.Value...)
 	return encodedBytes
 }
+
+func intHash64(x uint64) uint64 {
+	x ^= x >> 33
+	x *= 0xff51afd7ed558ccd
+	x ^= x >> 33
+	x *= 0xc4ceb9fe1a85ec53
+	x ^= x >> 33
+	return x
+}
+
+type baseApproxCountDistinct struct {
+	baseAggFunc
+}
+
+const (
+	// The maximum degree of buffer size before the values are discarded
+	uniquesHashMaxSizeDegree uint8 = 17
+	// The maximum number of elements before the values are discarded
+	uniquesHashMaxSize = uint32(1) << (uniquesHashMaxSizeDegree - 1)
+	// Initial buffer size degree
+	uniquesHashSetInitialSizeDegree uint8 = 4
+	// The number of least significant bits used for thinning. The remaining high-order bits are used to determine the position in the hash table.
+	uniquesHashBitsForSkip = 32 - uniquesHashMaxSizeDegree
+)
+
+type approxCountDistinctHashValue uint32
+
+// partialResult4ApproxCountDistinct use `BJKST` algorithm to compute approximate result of count distinct.
+// According to an experimental survey http://www.vldb.org/pvldb/vol11/p499-harmouch.pdf, the error guarantee of BJKST
+// was even better than the theoretical lower bounds (i.e., relative error was always far less than 1%).
+// For the calculation state, it uses a sample of element hash values with a size up to uniquesHashMaxSize. Compared
+// with the widely known HyperLogLog algorithm, this algorithm is less effective in terms of accuracy and
+// memory consumption (even up to proportionality), but it is adaptive. This means that with fairly high accuracy, it
+// consumes less memory during simultaneous computation of cardinality for a large number of data sets whose cardinality
+// has power law distribution (i.e. in cases when most of the data sets are small).
+// This algorithm is also very accurate for data sets with small cardinality and very efficient on CPU.
+type partialResult4ApproxCountDistinct struct {
+	size       uint32 /// Number of elements.
+	sizeDegree uint8  /// The size of the table as a power of 2.
+	skipDegree uint8  /// Skip elements not divisible by 2 ^ skipDegree.
+	hasZero    bool   /// The hash table contains an element with a hash value of 0.
```

Comment:
Why we need a specified variable for zero? As I see, the algorithm could handle the zero itself.

Reply:
In hash table, zero means empty. So, we need `hasZero` for hashed value 0.

Reply:
OK👌